File: vc-verifier/js/src/CredentialsVerifier.js
import jws from 'jws';
import crypto from 'crypto';
import { Constants } from './constant/Constants.js';
import { getJwsSigningInput } from './utils/JwsSigningInput.js';
import { URDNA2015Canonicalizer } from './utils/URDNA2015Canonicalizer.js';
import axios from 'axios';
import {validate} from "./validator/LdpValidator.js";
import {Errors} from "./constant/ValidatorConstants.js";
import {isNotNullOrEmpty} from "./validator/Utils.js";
import {preProcessVerifiableCredential} from "./utils/CredentialUtils.js";


const getPublicKeyFromVerificationMethod = async (url) => {
    let response = await axios.get(url);
    response = response.data;
    const publicKeyPem = response['publicKeyPem'];
    const publicKeyObject = crypto.createPublicKey(
        {
            key: publicKeyPem,
            format: 'pem',
            type: 'spki'
        }
    );
    if (!publicKeyObject)
        throw new Error("Error while creating Public Key Object");
    return publicKeyObject;
}

const verifyCredentialSignature = (jwsHeaderAlgoName, publicKey, actualData, signature) => {
    let isVerified = false;
    if (jwsHeaderAlgoName === Constants.JWS_PS256_SIGN_ALGO_CONST) {
        isVerified = crypto.verify(
            'RSA-SHA256',
            actualData,
            {
                key: publicKey,
                padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
            },
            signature
        );
    }
    else if (jwsHeaderAlgoName === Constants.JWS_RS256_SIGN_ALGO_CONST) {
        isVerified = crypto.verify('RSA-SHA256', actualData, publicKey, signature);
    }
    else {
        throw new Error('Verification algorithm not supported');
    }
    return isVerified;
}

/**
 * @deprecated This method has been deprecated because it is not extensible for future use cases of supporting different VC format's verification
 * Please use verify(credentials, format) instead, which is designed for supporting different VC formats.
 * This method only supports LDP VC format
 */
export const verifyCredentials = async (credential) => {
    try {
        preProcessVerifiableCredential(credential);
        return await verifySignature(credential)
    } catch (error) {
        throw error;
    }
}

/**
 * Validates the provided credentials against the specified format and Verifies the Signature.
 *
 * @param {Object} credential - The credentials to verify, containing necessary fields.
 * @param {string} credentialFormat - The expected format for the credentials (e.g., 'ldp').
 * @returns {Promise<Object>} The result of the verification, including a status and any relevant messages.
 */
export const verify = async (credential, credentialFormat) => {
    try {
        const validationMessage = validate(credential)

        if(isNotNullOrEmpty(validationMessage) && (validationMessage !== Errors.ERROR_VC_EXPIRED)){
            verificationFailure(validationMessage)
        }
        const verificationResult = await verifySignature(credential)
        if(!verificationResult) {
            return verificationFailure(`${Errors.SIGNATURE_VERIFICATION_FAILED}`)
        }
        return verificationSuccess(validationMessage)

    } catch (error) {
        return verificationFailure(`${Errors.EXCEPTION_DURING_VERIFICATION}${error.message}`)
    }
}

const verifySignature = async (credential) => {
    const { signature, header } = jws.decode(credential.proof.jws);
    const decodedSignature = Buffer.from(signature, 'base64');
    const publicKeyObject = await getPublicKeyFromVerificationMethod(credential.proof.verificationMethod);
    const canonicalisedCredential = await URDNA2015Canonicalizer(credential);
    const inputData = getJwsSigningInput(header, canonicalisedCredential);
    return verifyCredentialSignature(header.alg, publicKeyObject, inputData, decodedSignature);
}

const verificationSuccess = (message) => ({
    verificationStatus: true,
    verificationMessage: message ? message: ""
});

const verificationFailure = (errorMessage) => ({
    verificationStatus: false,
    verificationMessage: errorMessage
});

File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/CredentialsVerifier.kt
package io.mosip.vercred.vcverifier

import io.mosip.vercred.vcverifier.constants.CredentialFormat
import io.mosip.vercred.vcverifier.constants.CredentialFormat.LDP_VC
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_VC_EXPIRED
import io.mosip.vercred.vcverifier.constants.CredentialVerifierConstants.ERROR_CODE_VERIFICATION_FAILED
import io.mosip.vercred.vcverifier.constants.CredentialVerifierConstants.ERROR_MESSAGE_VERIFICATION_FAILED
import io.mosip.vercred.vcverifier.constants.CredentialVerifierConstants.EXCEPTION_DURING_VERIFICATION
import io.mosip.vercred.vcverifier.credentialverifier.CredentialVerifierFactory
import io.mosip.vercred.vcverifier.credentialverifier.VerifiableCredential
import io.mosip.vercred.vcverifier.data.CacheEntry
import io.mosip.vercred.vcverifier.data.CredentialStatusResult
import io.mosip.vercred.vcverifier.data.CredentialVerificationSummary
import io.mosip.vercred.vcverifier.data.ValidationStatus
import io.mosip.vercred.vcverifier.data.VerificationResult
import io.mosip.vercred.vcverifier.utils.Util
import java.util.logging.Logger


class CredentialsVerifier {
    private val logger = Logger.getLogger(CredentialsVerifier::class.java.name)
    private val credentialVerifierFactory = CredentialVerifierFactory()

    /**
     * @deprecated This method has been deprecated because it is not extensible for future use cases of supporting different VC format's verification
     * Please use verify(credentials: String, format: CredentialFormat) instead, which is designed for supporting different VC formats.
     * This method only supports LDP VC format
     */
    @Deprecated("This method has been deprecated because it is not extensible for future use cases of supporting different VC format's verification")
    fun verifyCredentials(credentials: String?): Boolean {
        if (credentials == null) {
            logger.severe("Error - Input credential is null")
            throw RuntimeException("Input credential is null")
        }
        val credentialVerifier = credentialVerifierFactory.get(LDP_VC)
        val isVerified = credentialVerifier.verify(credentials)

        if (!isVerified) {
            logger.warning("Credential verification failed")
            return false
        }

        return true
    }

    fun verify(credential: String,
               credentialFormat: CredentialFormat,
               walletCache: MutableMap<String, CacheEntry>? = null,
               expiryTime: Long? = null): VerificationResult {
        if (walletCache != null) {
            Util.walletCache = walletCache
        }
        expiryTime?.let {
            Util.ttlMillis = it
        }
        val credentialVerifier = credentialVerifierFactory.get(credentialFormat)
        val validationStatus = credentialVerifier.validate(credential)
        if (validationStatus.validationMessage.isNotEmpty() && !validationStatus.validationErrorCode.contentEquals(
                ERROR_CODE_VC_EXPIRED
            )
        ) {
            return VerificationResult(
                false,
                validationStatus.validationMessage,
                validationStatus.validationErrorCode,
                Util.walletCache
            )
        }
        return try {
            val verifySignatureStatus = credentialVerifier.verify(credential)
            if (verifySignatureStatus) {
                return VerificationResult(
                    true,
                    validationStatus.validationMessage,
                    validationStatus.validationErrorCode,
                    Util.walletCache
                )
            }
            return VerificationResult(
                false,
                ERROR_MESSAGE_VERIFICATION_FAILED,
                ERROR_CODE_VERIFICATION_FAILED,
                Util.walletCache
            )

        } catch (e: Exception) {
            val errorCode = validationStatus.validationErrorCode.takeIf { !it.isNullOrEmpty() }
                ?: ERROR_CODE_VERIFICATION_FAILED
            VerificationResult(false, "$EXCEPTION_DURING_VERIFICATION${e.message}", errorCode,Util.walletCache)
        }
    }

    fun getCredentialStatus(
        credential: String,
        credentialFormat: CredentialFormat,
        statusPurposeList: List<String> = emptyList()
    ): List<CredentialStatusResult> {
        try {
            val credentialStatusArray =
                credentialVerifierFactory.get(credentialFormat)
                    .checkStatus(credential, statusPurposeList)
            return credentialStatusArray ?: emptyList()
        } catch (e: Exception) {
            logger.severe("Error occurred while checking credential status: ${e.message}")
            throw e
        }
    }

    fun verifyAndGetCredentialStatus(
        credential: String,
        credentialFormat: CredentialFormat,
        statusPurposeList: List<String> = emptyList()
    ): CredentialVerificationSummary {
        val verificationResult = verify(credential, credentialFormat)
        if (verificationResult.verificationStatus) {
            val statusResults = getCredentialStatus(credential, credentialFormat, statusPurposeList)
            return CredentialVerificationSummary(verificationResult, statusResults)
        }
        return CredentialVerificationSummary(verificationResult, emptyList())
    }
}